Dockerfile.template

```
@@ -37,6 +37,12 @@ ENV HAPROXY_VERSION {{ .version }}
 ENV HAPROXY_URL {{ .url }}
 ENV HAPROXY_SHA256 {{ .sha256 }}
 
+{{ if ([ "2.0", "2.2", "2.4", "2.6", "2.7", "2.8" ] | index(env.version)) then ( -}}
+ENV HAPROXY_LUA_VERSION 5.3
+{{ ) else ( -}}
+ENV HAPROXY_LUA_VERSION 5.4
+{{ ) end -}}
+
```

Nice!  Is there a good documentation page from upstream we could point to here to help justify these choices? :sweat_smile:

I think I'd also like to avoid creating another explicit environment variable that I don't think is particularly interesting to downstream users, so I think we should adjust this slightly to use a template function instead:

```suggestion
{{
	def lua:
		# TODO some useful upstream link here? 😅
		if ([ "2.0", "2.2", "2.4", "2.6", "2.7", "2.8" ] | index(env.version)) then
			"5.3"
		else
			"5.4"
		end
-}}
```

(and then it can be used as `{{ lua }}` later in the script in the couple places we need it)

---

Thanks for your quick-reply!

>Is there a good documentation page from upstream we could point to here to help justify these choices?

Well there are multiple motivations :smile: :
- haproxy offers Lua 5.4 support (https://github.com/haproxy/haproxy/commit/08ed98fd7963968de49593304fdd9234812845a4) since mid 2020
- Lua 5.3 is EOL since September 2020 (https://www.lua.org/versions.html#5.3)

So it's probably best to consider the 5.4 version as the default one from now on, do you agree?

>I think I'd also like to avoid creating another explicit environment variable that I don't think is particularly interesting to downstream users, so I think we should adjust this slightly to use a template function instead:

Of course, thanks for the suggestion is does seem clearer indeed, I didn't know about this as I'm not very familiar with the templating format

About the remaining files in the commit, is it correct to include the Dockerfile modifications from running `update.sh` or only the template file should be included? 

Thanks :)

---

Ah, interesting.  That makes me very seriously wonder if we should actually be bumping to 5.4 more aggressively (like, in all versions 2.3+), but I imagine the user impact of that is probably high?  I really don't know much about Lua in general or how it's integrated in HAProxy, sorry. :sweat_smile:

> About the remaining files in the commit, is it correct to include the Dockerfile modifications from running `update.sh` or only the template file should be included?

Yep, it's correct to include them!  However, you might have a more enjoyable inner loop if you only run `apply-templates.sh` because that won't go look for new versions too. :grin:

---

> if we should actually be bumping to 5.4 more aggressively (like, in all versions 2.3+)

Well, theoretically there would be no impact (5.3->5.4 upgrade should only expose new features and bugfixes and is not supposed to cause breakage in existing scripts).. but I'm never confident when it comes to changing something that's been stable for a long time, how do you feel about that?

Or perhaps we should only target edge versions (2.7+) to make the risk more acceptable.. I cannot make my mind actually and that's why I initially suggested the 2.9 since it is currently under development :smile: 

>Yep, it's correct to include them! However, you might have a more enjoyable inner loop if you only run apply-templates.sh because that won't go look for new versions too

Copy that, thanks for the info! I'll update the PR shortly once we decided which versions we should be targeting! :)

---

> 5.3->5.4 upgrade should only expose new features and bugfixes and is not supposed to cause breakage in existing scripts

It's not 100% accurate, there are some minor incompatibilities that could possibly affect existing scripts:
https://www.lua.org/manual/5.4/manual.html#8

Although being minor, if users don't expect **any** regression at all, then it's probably not safe to change defaults on older stable versions.. right?